escrow/escrow.go

```diff
@@ -1,267 +1,209 @@
-	storage         PaymentChannelStorage
-	incomeValidator IncomeValidator
-	blockchain      EscrowBlockchainApi
+// lockingPaymentChannelService implements PaymentChannelService interface
+// using locks around proxied service call to guarantee that only one payment
+// at time is applied to channel
+type lockingPaymentChannelService struct {
+	storage          PaymentChannelStorage
+	blockchainReader *BlockchainChannelReader
+	locker           Locker
+	validator        *ChannelPaymentValidator
 }
 
-// NewEscrowPaymentHandler returns instance of handler.PaymentHandler to validate
-// payments via MultiPartyEscrow contract.
-func NewEscrowPaymentHandler(
-	processor *blockchain.Processor,
+// NewPaymentChannelService returns instance of PaymentChannelService to work
+// with payments via MultiPartyEscrow contract.
+func NewPaymentChannelService(
 	storage PaymentChannelStorage,
-	incomeValidator IncomeValidator,
-	config *viper.Viper) handler.PaymentHandler {
+	blockchainReader *BlockchainChannelReader,
+	locker Locker,
+	channelPaymentValidator *ChannelPaymentValidator) PaymentChannelService {
 
-	return &escrowPaymentHandler{
-		config:          config,
-		storage:         storage,
-		incomeValidator: incomeValidator,
-		blockchain:      processor,
+	return &lockingPaymentChannelService{
+		storage:          storage,
+		blockchainReader: blockchainReader,
+		locker:           locker,
+		validator:        channelPaymentValidator,
 	}
 }
 
-type escrowPaymentType struct {
-	grpcContext  *handler.GrpcStreamContext
-	channelID    *big.Int
-	channelNonce *big.Int
-	amount       *big.Int
-	signature    []byte
-	channel      *PaymentChannelData
+func (h *lockingPaymentChannelService) PaymentChannel(key *PaymentChannelKey) (channel *PaymentChannelData, ok bool, err error) {
+	storageChannel, storageOk, err := h.storage.Get(key)
+	if err != nil {
+		return
+	}
+
+	blockchainChannel, blockchainOk, err := h.blockchainReader.GetChannelStateFromBlockchain(key)
+	if !storageOk {
+		return blockchainChannel, blockchainOk, err
+	}
+	if err != nil || !blockchainOk {
+		return storageChannel, storageOk, nil
+	}
+
+	return MergeStorageAndBlockchainChannelState(storageChannel, blockchainChannel), true, nil
 }
 
-func (p *escrowPaymentType) String() string {
-	return fmt.Sprintf("{grpcContext: %v, channelID: %v, channelNonce: %v, amount: %v, signature: %v, channel: %v}",
-		p.grpcContext, p.channelID, p.channelNonce, p.amount, blockchain.BytesToBase64(p.signature), p.channel)
+type claimImpl struct {
+	payment *Payment
+	lock    Lock
 }
 
-func (h *escrowPaymentHandler) Type() (typ string) {
-	return EscrowPaymentType
+func (claim *claimImpl) Payment() *Payment {
+	return claim.payment
 }
 
-func (h *escrowPaymentHandler) Payment(context *handler.GrpcStreamContext) (payment handler.Payment, err *status.Status) {
-	channelID, err := handler.GetBigInt(context.MD, PaymentChannelIDHeader)
+func (claim *claimImpl) Finish() (err error) {
+	err = claim.lock.Unlock()
```

Comment:
It was discussed to check where this lock is really necessary.

Reply:
Thank you for spotting it! Fixed it and https://github.com/singnet/snet-daemon/pull/128#discussion_r233364630 in https://github.com/singnet/snet-daemon/pull/128/commits/8f09533a7e46be8bb0670774642346b0d8e6cb59